wasm/jit/jit_value_location_amd64.go

```diff
@@ -0,0 +1,225 @@
+)
+
+type valueLocation struct {
+	valueType wazeroir.SignLessType
+	// Set to -1 if the value is stored in the memory stack.
+	register int16
+	// Set to conditionalRegisterStateUnset if the value is not on the conditional register.
+	conditionalRegister conditionalRegisterState
+	// This is the location of this value in the (virtual) stack,
+	// even though if .register != -1, the value is not written into memory yet.
+	stackPointer uint64
+}
+
+func (v *valueLocation) registerType() (t generalPurposeRegisterType) {
+	switch v.valueType {
+	case wazeroir.SignLessTypeI32, wazeroir.SignLessTypeI64:
+		t = gpTypeInt
+	case wazeroir.SignLessTypeF32, wazeroir.SignLessTypeF64:
+		t = gpTypeFloat
+	default:
+		panic("unreachable")
+	}
+	return
+}
+func (v *valueLocation) setValueType(t wazeroir.SignLessType) {
+	v.valueType = t
+}
+
+func (v *valueLocation) setRegister(reg int16) {
+	v.register = reg
+	v.conditionalRegister = conditionalRegisterStateUnset
+}
+
+func (v *valueLocation) onRegister() bool {
+	return v.register != -1 && v.conditionalRegister == conditionalRegisterStateUnset
+}
+
+func (v *valueLocation) onStack() bool {
+	return v.register == -1 && v.conditionalRegister == conditionalRegisterStateUnset
+}
+
+func (v *valueLocation) onConditionalRegister() bool {
+	return v.conditionalRegister != conditionalRegisterStateUnset
+}
+
+func newValueLocationStack() *valueLocationStack {
+	return &valueLocationStack{
+		usedRegisters: map[int16]struct{}{},
+	}
+}
+
+type valueLocationStack struct {
+	stack         []*valueLocation
+	sp            uint64
+	usedRegisters map[int16]struct{}
+}
+
+func (s *valueLocationStack) clone() *valueLocationStack {
+	ret := &valueLocationStack{}
+	ret.sp = s.sp
+	ret.usedRegisters = make(map[int16]struct{}, len(ret.usedRegisters))
+	for r := range s.usedRegisters {
+		ret.markRegisterUsed(r)
+	}
+	ret.stack = make([]*valueLocation, len(s.stack))
+	for i, v := range s.stack {
+		ret.stack[i] = &valueLocation{
+			valueType:           v.valueType,
+			conditionalRegister: v.conditionalRegister,
+			stackPointer:        v.stackPointer,
+			register:            v.register,
+		}
+	}
+	return ret
+}
+func (s *valueLocationStack) pushValueOnRegister(reg int16) (loc *valueLocation) {
+	loc = &valueLocation{register: reg, conditionalRegister: conditionalRegisterStateUnset}
+	s.markRegisterUsed(reg)
+	s.push(loc)
+	return
+}
+
+func (s *valueLocationStack) pushValueOnStack() (loc *valueLocation) {
+	loc = &valueLocation{register: -1, conditionalRegister: conditionalRegisterStateUnset}
+	s.push(loc)
+	return
+}
+
+func (s *valueLocationStack) pushValueOnConditionalRegister(state conditionalRegisterState) (loc *valueLocation) {
+	loc = &valueLocation{register: -1, conditionalRegister: state}
+	s.push(loc)
+	return
+}
+
+func (s *valueLocationStack) push(loc *valueLocation) {
+	loc.stackPointer = s.sp
+	if s.sp >= uint64(len(s.stack)) {
```

Comment:
nit the if >= has to do with growing, so a trailing comment of // then grow can help I think. supernit: s.sp++ can be outside the if/else right?

Reply:
https://github.com/tetratelabs/wazero/pull/60/commits/5fd13b36250f0fb937bfc3867b52f4848d7f0022